### ISSUE TITLE
retry pod egress iptables rule creation

### ISSUE THREAD

**Issue by ghouscht** [cloudnativelabs/kube-router#880]
Hi,
First of all thank you for your work on kube-router! Unfortunately we have seen an issue with pod egress traffic. Kube-router sometimes fails to apply the masquerade rule for pod egress traffic during startup. This is most likely due to kube-proxy holding the iptables xtables lock, see the log message below:

```
E0412 20:00:51.251863       1 network_routes_controller.go:176] Error enabling Pod egress: Failed to add iptables rule to masquerade outbound traffic from pods: running [/sbin/iptables -t nat -A POSTROUTING -m set --match-set kube-router-pod-subnets src -m set ! --match-set kube-router-pod-subnets dst -m set ! --match-set kube-router-node-ips dst -j MASQUERADE --wait]: exit status 4: iptables: Resource temporarily unavailable.
```

kube-router does not retry creation of the masquerade rule and simply continues to run "normally", which of course causes problems when you expect that pod ip's are masqueraded. Relevant code can be found here: https://github.com/cloudnativelabs/kube-router/blob/33724aac05a642c1f10d88b4df661153996ce877/pkg/controllers/routing/network_routes_controller.go#L172-L178

In my opinion this isn't really a good solution, I would either expect that kube-router fails to start if it's not able to configure the pod egress rule and let k8s restart the pod or to retry a failed creation of the rule later.

Happy to help here. Please let me know your opinion.
Regards, Thom

**Comment by aauren** [comment]
Thanks for your bug report!

This is a duplicate of #506 which is planned to be fixed for the 1.0 release: https://github.com/cloudnativelabs/kube-router/projects/6

kube-router plans to add proper xtables.lock handling to circumvent this issue and avoid conflicts with other iptables manipulators like kube-proxy.

**Comment by ghouscht** [comment]
> Thanks for your bug report!
> 
> This is a duplicate of #506 which is planned to be fixed for the 1.0 release: https://github.com/cloudnativelabs/kube-router/projects/6
> 
> kube-router plans to add proper xtables.lock handling to circumvent this issue and avoid conflicts with other iptables manipulators like kube-proxy.

Thanks for referencing the issue, indeed this seems to have the same root cause. 
Can you please clarify what you mean with `kube-router plans to add proper xtables.lock handling`? I think simply mounting the xtables.lock from the host (as for example kube-proxy does) should "resolve" this issue, or not? Looks like kube-router already uses the `--wait` option when executing `iptables` which means it is waiting to acquire the xtables lock. Or am I missing something here?

**Comment by murali-reddy** [comment]
> In my opinion this isn't really a good solution, I would either expect that kube-router fails to start if it's not able to configure the pod egress rule and let k8s restart the pod or to retry a failed creation of the rule later.

@ghouscht agree. It should have been a fatal error

> Can you please clarify what you mean with kube-router plans to add proper xtables.lock handling? I think simply mounting the xtables.lock from the host (as for example kube-proxy does) should "resolve" this issue, or not? 

In what combination are you using kube-router? If kube-router is being used all-in-one solution then very unlikely to run into ` iptables: Resource temporarily unavailable`, as with in the kube-router pod all the modules (proxy, network policy, pod networking) will hold the lock or wait.  If you are using kube-router along with kube-proxy then since kube-router is not holding host xlocks file, there is possible contention between kube-router and kube-proxy

Either way mounting the xtables.lock from the host is the right solution to be incorportated.

**Comment by aauren** [comment]
Closed via #884
